src/audio/kpb.c

```diff
@@ -1322,26 +1322,38 @@ static void kpb_reset_history_buffer(struct hb *buff)
 	} while (buff != first_buff);
 }
 
-static inline bool validate_host_params(size_t host_period_size,
-					size_t host_buffer_size,
-					size_t bytes_per_ms)
+static inline bool validate_host_params(struct comp_data *kpb,
+					size_t host_period_size,
+					size_t host_buffer_size)
 {
-	/* Check host period size sanity.
-	 * Here we check if host period size (which defines interval
-	 * time) will allow us to drain more data then the interval
-	 * takes - as only such condition guarantees draining will end.
-	 * The formula:
-	 *	drained_data_in_one_interval_ms > interval_break_ms
-	 * where:
-	 * drained_data_in_one_interval_ms = (host_period_size * 2) [ms]
-	 * interval_break_ms = host_period_size / bytes_per_ms [ms]
+	/* The aim of this function is to perform basic check of host params
+	 * and reject them if they won't allow for stable draining.
+	 * Note however that this is highly recommended for host buffer to
+	 * be of history buffer size. This will quarantee "safe" draining.
+	 * By safe we mean no XRUNs(host was unable to read data on time),
+	 * or lose of data due to host delayed read. The later condition
+	 * is very likely after wake up from power state like d0ix.
 	 */
 
-	if (host_period_size < bytes_per_ms || /* Out of control */
-	    host_period_size >= (host_buffer_size / 2)) /* XRUN */
-		return false;
+	size_t kpb_hb_size = KPB_MAX_BUFFER_SIZE(kpb->config.sampling_width);
 
-	return true;
+	if (host_period_size == 0 || host_buffer_size == 0) {
+		/* Wrong host params */
+		return false;
+	} else if (host_buffer_size > kpb_hb_size) {
```

Comment:
@keyonjie, right, to be honest this part doesn't care about wake up time but only regular XRUN. 
Anyway, since we have confirmation that client will use four second buffer I think we can force such buffer size in this validation. Do you agree? This will free use from many problems.